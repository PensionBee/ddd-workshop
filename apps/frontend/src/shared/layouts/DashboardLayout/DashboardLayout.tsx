import type { ReactNode } from "react";
import { FooterContactSection } from "@shared/panels/FooterContactSection";
import { FooterCopyrightSection } from "@shared/panels/FooterCopyrightSection";
import { FooterCredentialsSection } from "@shared/panels/FooterCredentialsSection";
import { FooterFcaSection } from "@shared/panels/FooterFcaSection";
import { FooterQuestionSection } from "@shared/panels/FooterQuestionSection";
import { FooterSitemap } from "@shared/panels/FooterSitemap";
import { FooterSocialMediaList } from "@shared/panels/FooterSocialMediaList";
import { HeaderBeehiveMenu } from "@shared/panels/HeaderBeehiveMenu";
import { HeaderMainNavbar } from "@shared/panels/HeaderMainNavbar";

export type DashboardProps = {
  children?: ReactNode;
};

export const DashboardLayout: React.FC<DashboardProps> = ({
  children,
  ...props
}) => {
  return (
    <>
      <header>
        <HeaderMainNavbar />
        <HeaderBeehiveMenu />
      </header>
      <main data-layout="DashboardLayout" {...props}>
        {children}
      </main>
      <footer>
        <FooterQuestionSection />
        <FooterFcaSection />
        <FooterSocialMediaList />
        <FooterSitemap />
        <FooterContactSection />
        <FooterCopyrightSection />
        <FooterCredentialsSection />
      </footer>
    </>
  );
};
