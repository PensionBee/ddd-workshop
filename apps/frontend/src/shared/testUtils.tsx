import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
