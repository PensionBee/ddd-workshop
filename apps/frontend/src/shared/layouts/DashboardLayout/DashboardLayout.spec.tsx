import { render, vi } from "@shared/testUtils";
import { DashboardLayout } from "./DashboardLayout";

vi.mock("@shared/panels/HeaderBeehiveMenu", () => ({
  HeaderBeehiveMenu: () => <div data-testid="HeaderBeehiveMenu" />,
}));
vi.mock("@shared/panels/HeaderMainNavbar", () => ({
  HeaderMainNavbar: () => <div data-testid="HeaderMainNavbar" />,
}));
vi.mock("@shared/panels/FooterContactSection", () => ({
  FooterContactSection: () => <div data-testid="FooterContactSection" />,
}));
vi.mock("@shared/panels/FooterCopyrightSection", () => ({
  FooterCopyrightSection: () => <div data-testid="FooterCopyrightSection" />,
}));
vi.mock("@shared/panels/FooterCredentialsSection", () => ({
  FooterCredentialsSection: () => (
    <div data-testid="FooterCredentialsSection" />
  ),
}));
vi.mock("@shared/panels/FooterFcaSection", () => ({
  FooterFcaSection: () => <div data-testid="FooterFcaSection" />,
}));
vi.mock("@shared/panels/FooterQuestionSection", () => ({
  FooterQuestionSection: () => <div data-testid="FooterQuestionSection" />,
}));
vi.mock("@shared/panels/FooterSitemap", () => ({
  FooterSitemap: () => <div data-testid="FooterSitemap" />,
}));
vi.mock("@shared/panels/FooterSocialMediaList", () => ({
  FooterSocialMediaList: () => <div data-testid="FooterSocialMediaList" />,
}));

describe("DashboardLayout", () => {
  const TEST_ID = "DashboardLayout";
  it("should render", () => {
    const { getByTestId, asFragment } = render(
      <DashboardLayout data-testid={TEST_ID} />
    );
    expect(getByTestId(TEST_ID)).toBeInTheDocument();
    expect(asFragment).toMatchSnapshot();
  });

  it("should render each panel", () => {
    const { getByTestId } = render(<DashboardLayout />);
    expect(getByTestId("HeaderBeehiveMenu")).toBeInTheDocument();
    expect(getByTestId("FooterContactSection")).toBeInTheDocument();
    expect(getByTestId("FooterCopyrightSection")).toBeInTheDocument();
    expect(getByTestId("FooterCredentialsSection")).toBeInTheDocument();
    expect(getByTestId("FooterFcaSection")).toBeInTheDocument();
    expect(getByTestId("FooterQuestionSection")).toBeInTheDocument();
    expect(getByTestId("FooterSitemap")).toBeInTheDocument();
    expect(getByTestId("FooterSocialMediaList")).toBeInTheDocument();
    expect(getByTestId("HeaderMainNavbar")).toBeInTheDocument();
  });
});
