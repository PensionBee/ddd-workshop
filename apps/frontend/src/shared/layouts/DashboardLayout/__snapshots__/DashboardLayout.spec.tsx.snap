// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DashboardLayout > should render 1`] = `[Function]`;

exports[`DashboardLayout should render 1`] = `[Function]`;
